src/Data/Text/IO.hs

```diff
@@ -176,12 +179,14 @@ hGetLine = hGetLineWith T.concat
 hPutStr :: Handle -> Text -> IO ()
 -- This function is lifted almost verbatim from GHC.IO.Handle.Text.
 hPutStr h t = do
-  (buffer_mode, nl) <-
+  (buffer_mode, nl, isUTF8) <-
        wantWritableHandle "hPutStr" h $ \h_ -> do
                      bmode <- getSpareBuffer h_
-                     return (bmode, haOutputNL h_)
+                     let isUTF8 = maybe False ((== "UTF-8") . textEncodingName) $ haCodec h_
```

Comment:
Let's use `textEncodingName GHC.IO.Encoding.utf8` instead of hard-coded `"UTF-8"`.

Do we have any idea of performance impact? Comparing strings is not cheap. Maybe we can use `GHC.Exts.unsafePtrEquality# GHC.IO.Encoding.utf8 (haCodec h_)`. After all, it's just an optimization.

Reply:
Thanks for finding this the string comparison was hurting performance a lot when I benchmarked this.

Reply:
One problem is it is possible to make your own utf-8 encoding https://hackage.haskell.org/package/base-4.20.0.0/docs/GHC-IO-Encoding.html#t:TextEncoding

Reply:
But who does that?

Reply:
If you crafted your own Unicode `TextEncoding`, you would not benefit from optimized code path, but that's fine.

It's actually safer to check `GHC.Exts.unsafePtrEquality# GHC.IO.Encoding.utf8 (haCodec h_)` than `== "UTF-8"`, because everyone can set `textEncodingName` to whatever they like.

Reply:
I have tried implementing this but I am not familiar with pointer equality. If I implemented this correctly, please resolve. Otherwise, I'll try and resolve the errors.